### PR TITLE
Add SISCORE dataset.

### PR DESCRIPTION
**Pull request by jessicayung**
([Dataset homepage](https://github.com/google-research/si-score))

# Task Description

As we begin to bring image classification models into the real world, it is important that they are robust and do not overfit to academic datasets. While robustness is hard to define, one can think of intuitive factors that image classification models should be robust to, such as changes in object location or differing weather conditions. Models should at least be robust to these factors. To test this, ideally we would have datasets that vary only these factors. For some factors such as image compression and differing weather conditions, datasets such as Imagenet-C already exist, whereas for common-sense factors such as object size and object location, such datasets did not yet exist prior to SI-SCORE.

In SI-SCORE, we take objects and backgrounds and systematically vary object size, location and rotation angle so we can study the effect of changing these factors on model performance. We also provide the code, object images and background images and hope that researchers can create their own datasets to test robustness to other factors of variation.

Here are some sample images from the dataset:

  

  

## Evaluation Metrics
The dataset uses accuracy as an evaluation metric. One can also calculate the accuracy per e.g. object size to compare model performance across different object sizes (/locations of the object in the image / object rotation angles).

## Expected Insights/Relevance
Discovering differences in robustness to changes in object size/location/rotation angle between image classification models, such as ResNet-50s with GroupNorm being better at classifying small objects than ResNet-50s with BatchNorm. (See [our workshop paper](https://arxiv.org/abs/2104.04191) for more details.)

Testing hypotheses such as whether Vision Transformer models perform worse when objects are on patch intersections.

# Access

## Data

[Dataset homepage](https://github.com/google-research/si-score)
Big zip files: `https://s3.us-east-1.amazonaws.com/si-score-dataset/{rotation|location|area}.zip`

## License
Yes - Apache 2.0 license.

### CODE REVIEW

**Review by stes — CHANGES_REQUESTED** [review]
Hi @jessicayung , thanks a lot for your contribution. We are working towards the first release of the workshop and for this, released our [contribution guide](https://github.com/shift-happens-benchmark/icml-2022/blob/main/CONTRIBUTING.rst) with the aim of unifying the implementation of all contributions.

Please have a look, the required changes should be relatively small and are mostly concerns with placing the files within the package, and adding documentation.

From reviewing your PR, the main required changes are:

- [ ] Place the files into their own package inside the `tasks/` folder
- [ ] Add a README and optionally citation info
- [ ] We are also looking into the best ways to perform unit or integration tests and would be happy to get some feedback on how this would be most effective for you benchmark. Would it be possible to add a "dry-run/demo" option to the benchmark, and e.g. run it only on a small subset of the data? No immediate action/update is required right now, we are first gathering opinions from all authors on this.

Thanks a lot for your contribution, and please ping me if you have any questions. If you have technical questions, also feel free to ping @kotekjedi who is happy to help with implementation details.

**Comment by stes** [comment]
Hi!

This is an automated report for the documentation coverage of your PR. To provide good user experience for the benchmark, we aim to cover most functions with doc strings, using the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

We would greatly appreciate your help in assuring that docstring coverage is sufficiently high for your contribution.

The current coverage is:
| Name                         |     Total |     Miss |     Cover |     Cover% |
|------------------------------|-----------|----------|-----------|------------|
| models/torchvision.py        |         4 |        0 |         4 |       100% |
| tasks/siscore.py             |         8 |        5 |         3 |        38% |
| tasks/siscore_labels.py      |         1 |        0 |         1 |       100% |
|------------------------------|-----------|----------|-----------|------------|
| TOTAL                        |        13 |        5 |         8 |      61.5% |
RESULT: FAILED (minimum: 80.0%, actual: 61.5%)

You can reproduce this result by running

```
interrogate -v --ignore-nested-functions --ignore-private --ignore-semiprivate --ignore-init-method
```

on the files you added to your PR.

**Comment by stes** [comment]
@kotekjedi could you run the pre-commit hooks on this PR so that we pass the first test?

File: shifthappens/tasks/siscore/siscore.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+from shifthappens.data.base import DataLoader
+from shifthappens.models import base as sh_models
+from shifthappens.models.base import PredictionTargets
+from shifthappens.tasks.base import abstract_variable
+from shifthappens.tasks.base import Task
+from shifthappens.tasks.base import variable
+from shifthappens.tasks.metrics import Metric
+from shifthappens.tasks.task_result import TaskResult
+
+import siscore_labels
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:33]:
```suggestion
```

File: shifthappens/tasks/siscore/siscore.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+"""
+
+import dataclasses
+import os
+from typing import Any
+from typing import Callable
+from typing import Optional
+from typing import Tuple
+
+import numpy as np
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:16]:
```suggestion
import numpy as np
import siscore_labels
```

**Review by kotekjedi — COMMENTED** [review]
Changes to pass `isort` test

File: shifthappens/tasks/siscore/siscore.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        ):
+            all_predicted_labels_list.append(predictions.class_labels)
+        all_predicted_labels = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = (
+            all_predicted_labels
+            == np.array(self.ch_dataset.targets)[: len(all_predicted_labels)]
+        )
+
+        # TODO(jessicayung): return more fine-grained metric as well,
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:141]: Should we remove this TODO?

File: shifthappens/tasks/siscore/siscore.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            accuracy=accuracy,
+            summary_metrics={Metric.Robustness: "accuracy"},
+        )
+
+
+# Varying object size.
+@sh_benchmark.register_task(
+    name="SISCORE (Size)",
+    relative_data_folder="siscore",
+    standalone=False,
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:155]:
```suggestion
    standalone=True,
```

File: shifthappens/tasks/siscore/siscore.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+@dataclasses.dataclass
+class SISCORESize(SISCOREVariantBase):
+    resource: Tuple[str, ...] = variable(("size"))
+
+
+# Varying object rotation angle.
+@sh_benchmark.register_task(
+    name="SISCORE (Rotation)",
+    relative_data_folder="siscore",
+    standalone=False,
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:166]:
```suggestion
    standalone=True,
```

File: shifthappens/tasks/siscore/siscore.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+@dataclasses.dataclass
+class SISCORERotation(SISCOREVariantBase):
+    resource: Tuple[str, ...] = variable(("rotation"))
+
+
+# Varying object location.
+@sh_benchmark.register_task(
+    name="SISCORE (Location)",
+    relative_data_folder="siscore",
+    standalone=False,
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:177]:
```suggestion
    standalone=True,
```

File: shifthappens/tasks/siscore/siscore.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+import dataclasses
+import os
+from typing import Any
+from typing import Callable
+from typing import Optional
+from typing import Tuple
+
+import numpy as np
+import siscore_labels
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:17]:
```suggestion
```

File: shifthappens/tasks/siscore/siscore.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+# Varying object rotation angle.
+@sh_benchmark.register_task(
+    name="SISCORE (Rotation)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORERotation(SISCOREVariantBase):
+    resource: Tuple[str, ...] = variable(("rotation"))
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:169]:
```suggestion
class SISCORERotation(SISCOREVariantBase):
    """Various rotations subset."""
    
    resource: Tuple[str, ...] = variable(("rotation"))
```

File: shifthappens/tasks/siscore/siscore.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+# Varying object size.
+@sh_benchmark.register_task(
+    name="SISCORE (Size)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORESize(SISCOREVariantBase):
+    """Various object size subset."""
+
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:159]:
```suggestion

```

File: shifthappens/tasks/siscore/siscore.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+# Varying object rotation angle.
+@sh_benchmark.register_task(
+    name="SISCORE (Rotation)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORERotation(SISCOREVariantBase):
+    """Various rotations subset."""
+
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:172]:
```suggestion

```

File: shifthappens/tasks/siscore/siscore.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+# Varying object location.
+@sh_benchmark.register_task(
+    name="SISCORE (Location)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORELocation(SISCOREVariantBase):
+    """Various locations subset."""
+
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:185]:
```suggestion

```

File: shifthappens/tasks/siscore/siscore.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+# Varying object size.
+@sh_benchmark.register_task(
+    name="SISCORE (Size)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORESize(SISCOREVariantBase):
+    """Various object size subset."""
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:158]:
```suggestion
    """Various object size subset."""

```

File: shifthappens/tasks/siscore/siscore.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+# Varying object rotation angle.
+@sh_benchmark.register_task(
+    name="SISCORE (Rotation)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORERotation(SISCOREVariantBase):
+    """Various rotations subset."""
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:170]:
```suggestion
    """Various rotations subset."""

```

File: shifthappens/tasks/siscore/siscore.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+# Varying object location.
+@sh_benchmark.register_task(
+    name="SISCORE (Location)",
+    relative_data_folder="siscore",
+    standalone=True,
+)
+@dataclasses.dataclass
+class SISCORELocation(SISCOREVariantBase):
+    """Various locations subset."""
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:182]:
```suggestion
    """Various locations subset."""

```

File: shifthappens/tasks/siscore/siscore.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+            all_predicted_labels_list.append(predictions.class_labels)
+        all_predicted_labels = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = (
+            all_predicted_labels
+            == np.array(self.ch_dataset.targets)[: len(all_predicted_labels)]
+        ).mean()
+
+        # TODO(jessicayung): return more fine-grained metric as well,
+        # i.e. accuracy per object size/rotation angle etc.
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:141]:
```suggestion
        
```

File: shifthappens/tasks/siscore/siscore.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+        ):
+            all_predicted_labels_list.append(predictions.class_labels)
+        all_predicted_labels = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = (
+            all_predicted_labels
+            == np.array(self.ch_dataset.targets)[: len(all_predicted_labels)]
+        ).mean()
+
+        # TODO(jessicayung): return more fine-grained metric as well,
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:140]:
```suggestion
```

File: shifthappens/tasks/siscore/siscore.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+            dataloader, PredictionTargets(class_labels=True)
+        ):
+            all_predicted_labels_list.append(predictions.class_labels)
+        all_predicted_labels = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = (
+            all_predicted_labels
+            == np.array(self.ch_dataset.targets)[: len(all_predicted_labels)]
+        ).mean()
+
```
Comment (kotekjedi) [inline shifthappens/tasks/siscore/siscore.py:139]:
```suggestion
```

**Review by zimmerrol — APPROVED** [review]
